### PR TITLE
OKE addon support

### LINKED ISSUE #254: Add support for OKE Addons API
**What would you like to be added**:

Add support for OKE Addons API

**Why is this needed**:

### PR DESCRIPTION
**Pull request by shyamradhakrishnan**
**What this PR does / why we need it**:

Add support for OKE Addons

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #254

### CODE REVIEW

**Comment by shyamradhakrishnan** [comment]
unit test
```
?   	github.com/oracle/cluster-api-provider-oci	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta1	41.592s	coverage: 24.1% of statements
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta2	0.042s	coverage: 15.8% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/config	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/metrics	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/ociutil	0.005s	coverage: 19.6% of statements
ok  	github.com/oracle/cluster-api-provider-oci/cloud/scope	240.073s	coverage: 75.6% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer/mock_lb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/util	0.165s	coverage: 76.1% of statements
ok  	github.com/oracle/cluster-api-provider-oci/controllers	26.104s	coverage: 67.5% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1	27.917s	coverage: 18.8% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta2	0.024s	coverage: 8.8% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/controllers	1.058s	coverage: 56.7% of statements
?   	github.com/oracle/cluster-api-provider-oci/feature	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/version	[no test files]
go tool cover -func=coverage.out -o coverage.txt
go tool cover -html=coverage.out -o coverage.html
```

**Comment by shyamradhakrishnan** [comment]
e2e
```
[SynchronizedAfterSuite] PASSED [0.888 seconds]
[SynchronizedAfterSuite]
/home/ubuntu/go/src/github.com/oracle/cluster-api-provider-oci/test/e2e/e2e_suite_test.go:256

  Timeline >>
  STEP: Tearing down the management cluster @ 06/22/23 07:05:08.565
  << Timeline
------------------------------

Ran 7 of 26 Specs in 3797.628 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 19 Skipped

```

File: cloud/scope/managed_control_plane.go [hunk 1]
```diff
@@ -669,6 +671,174 @@ func (s *ManagedControlPlaneScope) getSpecFromActual(cluster *oke.Cluster) *infr
...
+}
+
+func (s *ManagedControlPlaneScope) handleDeletedAddon(ctx context.Context, okeCluster *oke.Cluster, addonName string) error {
+	resp, err := s.ContainerEngineClient.GetAddon(ctx, oke.GetAddonRequest{
+		ClusterId: okeCluster.Id,
+		AddonName: common.String(addonName),
+	})
+	if err != nil {
+		if ociutil.IsNotFound(err) {
+			s.OCIManagedControlPlane.RemoveAddonStatus(addonName)
```
Comment (joekr) [inline cloud/scope/managed_control_plane.go:782]: Should we return here? won't this continue down the rest of the logic so the `resp` could be nil on line 787?

File: cloud/scope/managed_control_plane.go [hunk 2]
```diff
@@ -669,6 +671,174 @@ func (s *ManagedControlPlaneScope) getSpecFromActual(cluster *oke.Cluster) *infr
...
+		} else {
+			return err
+		}
+	}
+	addonState := resp.LifecycleState
+	// if the addon is not in deleting state, call delete
+	if !(addonState == oke.AddonLifecycleStateDeleting) {
+		s.Info(fmt.Sprintf("Deleting addon %s", addonName))
+		_, err := s.ContainerEngineClient.DisableAddon(ctx, oke.DisableAddonRequest{
```
Comment (joekr) [inline cloud/scope/managed_control_plane.go:791]: I don't fully follow: if it is in deleting state why is a "disable" call needed? Do you mind commenting on why this is needed?

File: cloud/scope/managed_control_plane.go [hunk 3]
```diff
@@ -669,6 +671,174 @@ func (s *ManagedControlPlaneScope) getSpecFromActual(cluster *oke.Cluster) *infr
...
+			Value: c.Value,
+		}
+	}
+	return config
+}
+
+func getAddon(addons []infrav2exp.Addon, name string) *infrav2exp.Addon {
+	for _, addon := range addons {
+		if *addon.Name == name {
+			return &addon
```
Comment (joekr) [inline cloud/scope/managed_control_plane.go:836]: I don't recall: do we need to do the loop variable capture stuff here? or is that only for `append` to an array?

**Review by joekr — CHANGES_REQUESTED** [review]
super small stuff. I think all of this makes sense to me.

File: test/e2e/e2e_suite_test.go [hunk 4]
```diff
@@ -24,6 +24,7 @@ import (
 	"encoding/base64"
 	"flag"
 	"fmt"
+	"github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine"
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply test/e2e/e2e_suite_test.go]: fixed

File: cloud/scope/managed_control_plane.go (hunk 1 above)
Comment (shyamradhakrishnan) [reply cloud/scope/managed_control_plane.go:786]: yep, good point. added a unit test also for this.

File: cloud/scope/managed_control_plane.go (hunk 3 above)
Comment (shyamradhakrishnan) [reply cloud/scope/managed_control_plane.go]: great point, fixed.

File: cloud/scope/managed_control_plane.go (hunk 2 above)
Comment (shyamradhakrishnan) [reply cloud/scope/managed_control_plane.go:806]: changed the code a bit, so disable is the call to delete an addon with the flag. if the flag is in deleting/deleted state, nothing to do. Else call delete. Hope it is better now.

**Review by joekr — APPROVED** [review]

**Pull request merged**